docs/csharp/roslyn-sdk/tutorials/how-to-write-csharp-analyzer-code-fix.md

```diff
@@ -0,0 +1,505 @@
+
+For efficiency, the first step is to refactor the two tests into data driven tests. Then, you only need to define a couple string constants for each new test. While your refactoring, rename both methods to better names. Replace `Test1` with this test that ensures no diagnostic is raised:
+
+```csharp
+[DataTestMethod]
+[DataRow("")]
+public void WhenTestCodeIsValidNoDiagnosticIsTriggered(string testCode)
+{
+    VerifyCSharpDiagnostic(testCode);
+}
+```
+
+You can create a new data row for this test by defining any code fragment that should not cause your diagnostic to trigger a warning. This overload of `VerifyCSharpDiagnostic` passes when there are no diagnostics triggered for the source code fragment.  
+
+Next, replace `Test2` with this test that ensures a diagnostic is raised and a code fix applied for the source code fragment:
+
+```csharp
+[DataTestMethod]
+[DataRow(LocalIntCouldBeConstant, LocalIntCouldBeConstantFixed, 10, 13)]
+public void WhenDiagosticIsRaisedFixUpdatesCode(
+    string test,
+    string fixTest,
+    int line,
+    int column)
+{
+    var expected = new DiagnosticResult
+    {
+        Id = MakeConstAnalyzer.DiagnosticId,
+        Message = new LocalizableResourceString(nameof(MakeConst.Resources.AnalyzerMessageFormat), MakeConst.Resources.ResourceManager, typeof(MakeConst.Resources)).ToString(),
+        Severity = DiagnosticSeverity.Warning,
+        Locations =
+            new[] {
+                    new DiagnosticResultLocation("Test0.cs", line, column)
+                }
+    };
+
+    VerifyCSharpDiagnostic(test, expected);
+
+    VerifyCSharpFix(test, fixTest);
+}
+```
+
+The preceding code also made a couple changes to the code that builds the expected diagnostic result. It uses the public constants registered in the `MakeConst` analyzer. In addition, it uses two string constants for the input and fixed source:
+
+[!code-csharp[string constants for fix test](~/samples/csharp/roslyn-sdk/Tutorials/MakeConst/MakeConst.Test/MakeConstUnitTests.cs#FirstFixTest "string constants for fix test")]
+
+Run these two tests to make sure they pass. In Visual Studio, open the **Test Explorer** by selecting **Test** > **Windows** > **Test Explorer**.  The press the **Run All** link.
+
+## Create tests for valid declarations
+
+As a general rule, analyzers should exist as quickly as possible, doing minimal work. Visual Studio calls registered analyzers as the user edits code. Responsiveness is a key requirement.  There are a number of tests to add for declarations that should not raise your diagnostic. There are several test cases for code that should not raise your diagnostic. Your analyzer already handles one of those tests, the case where a variable is assigned after being initialized. Add the following string constant to your tests to represent that case:
```

Comment:
Did you mean "As a general rule, analyzers should **exist** as quickly as possible" ? Or did you mean **exit**

Reply:
exit